openssh/plan.sh

```diff
@@ -29,4 +29,5 @@ do_install() {
   make install-nosysconf
   mkdir -p "${pkg_prefix}/var/empty"
   chmod 700 "${pkg_prefix}/var/empty"
+  chattr +i "${pkg_prefix}/var/empty"
```

Comment:
This builds fine, however when I go to remove my studio I get the following error message when using a linux chroot studio: 

```
hab studio rm
   hab-studio: Destroying Studio at /hab/studios/home--ubuntu--core-plans (default)
rm: can't remove '/hab/studios/home--ubuntu--core-plans/hab/pkgs/core/openssh/7.5p1/20181129161359/var/empty': Operation not permitted
```

which is going to cause problems on builder if this is merged.  

I'd like to help you solve this since it's blocking your gitea package, but I'm having difficulty replicating the issue.  I've tried `hab svc load core/openssh` in the docker studio on MacOS, and in an Ubuntu 18.10 vm and CentOS7 w/ selinux enabled and they all started successfully.   Can you share information about the environment you saw this in (Docker/LinuxDistro/Filesystem/etc.)?  Maybe it's an interaction with starting it from another packages hooks?

Reply:
If I open up a habitat studio, it loads fine. The issue is loading the service in another plan. I'm on Arch Linux.

The relevant part of the run hook looks like this:
```
#!/bin/sh

{{ ~#if cfg.enable_openssh }}
hab svc load core/openssh
{{ ~/if }}
```

The error message when setting `enable_openssh=true` is:
```
openssh.default(O): /hab/pkgs/core/openssh/7.5p1/20180608142228/var/empty must be owned by root and not group or world-writable.
```

Reply:
A hacky-scratchy solution would be to add a `chmod 700` to the run hook, which I've tested and works. But it's very confusing as to why using it within another plan changes the permissions.

Reply:
I've still been unable to replicate this. I've generated an empty plan that has a run hook that loads openssh, and then sleeps, and it loaded successfully. 

Correct me if I'm wrong,  but it looks like your intention with this is to encode that a running openssh service is a dependency of the gitea service,  and that opensssh service can optionally be provided by Habitat.  

I'm not certain that loading in the hooks is necessarily the right approach for this, though it's not wrong and _should_ work.  

Does Gitea need any information from the openssh service, such as port?   It may enough to declare an [optional bind ](https://www.habitat.sh/docs/developing-packages/#the-difference-between-pkg_binds-and-pkg_binds_optional) on openssh, though there's no guarantee that you'd get the openssh service on localhost if you were running a large ring with multiple openssh services in the same service group. 

[Composite Plans](https://www.habitat.sh/blog/2018/10/shelving-composites/), when available, would also be a potential way to describe this relationship between services but there's not currently a timeline on when we'll be revisiting those. 

I _think_ what you'll probably want to do here is make a documentation note that this requires openssh to be running on the same host, and leave it up to the provisioning layer to set that up either by utilizing the hab package, or the native distro packages. 

Does that track with what you're trying to do?

Reply:
Yep, you've got it. In the config file for gitea (app.ini), I added an extra option that maps to `default.toml` for `ssh_listen_host`, so anyone can specify where the ssh server is. It's a great addition, because it adds extra modularity.

I added this block to the `docker-compose.yml` that I'm currently working with:
```
  sshd:
    image: core/openssh:latest
    restart: always
    networks:
      - buildbot
    ports:
      - "2222:22"
```

And I have this block for the gitea host:
```
    environment:
      HAB_GITEA: |
        ssh_listen_host = "sshd"
```

I did a fresh export of `core/openssh` with `hab pkg export docker core/openssh`.
The problem still remains though:
```
sshd_1_6648bc60f610 | openssh.gitea(SV): Starting service as user=root, group=root
sshd_1_6648bc60f610 | openssh.gitea(O): /hab/pkgs/core/openssh/7.5p1/20180608142228/var/empty must be owned by root and not group or world-writable.
```

That's running as its own host, not inside of the gitea container, within the compose cluster.

Reply:
Also, with the `chattr` modified hab studio, I could run `hab studio rm` cleanly.

Reply:
Something really odd is going on. I built and exported the plan with `chattr +i`. It still has the permissions as `drwxrwx---    2 root     root          4096 Jun  8 14:23 empty`. I am absolutely baffled by this.

Reply:
Looking closer at this, I agree with @smacfarlane - In your testing, I would have an optional bind to openssh, which you would start separately. OpenSSH should not start as a result of starting your gitea service. They depend on each other, and thats fine, but start them separately.

This will ensure you're using the correct user to start the service, and there is no contamination from it.

Reply:
@TheDarkula after doing some more reading and testing, I don't think we can accept this PR.  From the [man page](https://linux.die.net/man/1/chattr):

>A file with the 'i' attribute cannot be modified: it cannot be deleted or renamed, no link can be created to this file and no data can be written to the file. Only the superuser or a process possessing the CAP_LINUX_IMMUTABLE capability can set or clear this attribute. 

Studios need the ability to delete files when they're removed, and supervisors will need this as well when they gain the ability to prune old services.  Additionally, the Docker studio doesn't have that capability by default, and even launching the Docker studio (on mac) with that capability results in the following error message: `chattr: Inappropriate ioctl for device while reading flags on /hab/pkgs/smacfarlane/openssh/7.5p1/20181203181037/var/empty`

Thank you for your contribution and working through this with us. I'm going to leave this open a little longer if there's an alternative solution you want to look at, but otherwise will be closing this.

Reply:
That's all good. I'm still working on figuring out what's changing the permissions on /var/empty, because it happens even if I do `hab pkg export docker core/openssh`.